### PR TITLE
Cf implement get desired state

### PR DESCRIPTION
**Pull request by fgimenez**
Towards https://github.com/giantswarm/giantswarm/issues/2023

It also includes the separation of methods in files. The current ASG template is copied to a string variable in `service/resources/cloudformation/template.go`, still not used. I've removed the parameters from there but I'm afraid we are going to need them (plus outputs too) in order to be able to compare states. We can discuss this in the upcoming `GetCreateState` branch.

### CODE REVIEW

File: service/key/key.go [hunk 1]
```diff
@@ -88,3 +89,23 @@ func WorkerInstanceType(customObject awstpr.CustomObject) string {
...
+	customObjectPointer, ok := v.(*awstpr.CustomObject)
+	if !ok {
+		return awstpr.CustomObject{}, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", &awstpr.CustomObject{}, v)
+	}
+	customObject := *customObjectPointer
+
+	return customObject, nil
+}
+
+func MainStackName(customObject awstpr.CustomObject) string {
```
Comment (xh3b4sd) [inline service/key/key.go]: FYI we tried to order methods alphabetically in the past.

File: service/resource/cloudformation/delete.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	"context"
+
+	"github.com/giantswarm/operatorkit/framework"
+)
+
+func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error {
+	return nil
+}
+
+func (r *Resource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
```
Comment (xh3b4sd) [inline service/resource/cloudformation/delete.go]: Would have been cool to do the file separation in a separate PR. Reviews would way be faster on such "dumb" changes as opposed to mix it with more complex logical changes.

File: service/resource/cloudformation/template.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    UpdatePolicy:
+      AutoScalingRollingUpdate:
+        # minimum amount of instances that must always be running during a rolling update
+        MinInstancesInService: 2
+        # only do a rolling update of this amount of instances max
+        MaxBatchSize: 2
+        # after creating a new instance, pause operations on the ASG for this amount of time
+        PauseTime: PT10S
+
+`
```
Comment (rossf7) [inline service/resource/cloudformation/template.go]: We want to start dev with the existing workers ASG template. But I think we should have a single main template that includes a child "control plane" template.

That way there is a single stack to update / delete and we can extend it later with more child templates. I think this can come in later PRs.

File: service/resource/cloudformation/desired.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+		return nil, microerror.Mask(err)
+	}
+
+	mainStack, err := newMainStack(customObject)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return mainStack, nil
+}
```
Comment (xh3b4sd) [inline service/resource/cloudformation/desired.go]: Some general questions to sort out where we are going to. In past implementations of operators we struggled with finding out what the actual state of a resource being reconciled should be. I want to prevent a lot of back and forth and get on the same page. One requirement for the framework is that the state being dispatched between methods should always be the same. Here we assume the desired state is `*awsCF.CreateStackInput`. Are we sure `*awsCF.CreateStackInput` is also what is returned by `GetCurrentState` and that `*awsCF.CreateStackInput` is what gets provided to `Apply*Change` methods? Do our dependency clients we rely on allow to fetch something as `*awsCF.CreateStackInput` or can we make up the current state? Do our dependency clients we rely on allow to update or delete state of this type?

File: service/resource/cloudformation/template.go (hunk 3 above)
Comment (fgimenez) [reply service/resource/cloudformation/template.go]: Makes perfect sense, thx!

File: service/resource/cloudformation/desired.go (hunk 4 above)
Comment (rossf7) [reply service/resource/cloudformation/desired.go]: Yes, good point the state object is very important. For the azure-operator we defined our own type for the Deployment which is pretty much equivalent to the Cloud Formation resource here.

@fgimenez Lets sync on this tomorrow.

File: service/key/key.go (hunk 1 above)
Comment (fgimenez) [reply service/key/key.go]: Sure, done

File: service/resource/cloudformation/desired.go (hunk 4 above)
Comment (fgimenez) [reply service/resource/cloudformation/desired.go]: > Are we sure *awsCF.CreateStackInput is also what is returned by GetCurrentState and that *awsCF.CreateStackInput is what gets provided to Apply*Change methods?

Nott at all! :) Thx a lot makes total sense, I've pushed changes to use a `Stack` type where we can put the required fields, pls take another look.

> @fgimenez Lets sync on this tomorrow.

@rossf7 sure, we need to talk too about how to compare states.

File: service/resource/cloudformation/resource.go [hunk 5]
```diff
@@ -1,72 +1,68 @@
...
+	logger    micrologger.Logger
 }
 
-// New creates a new configured config map resource.
+// StackState is the state representation pn which the resource methods work
+type StackState struct {
+	Name         string
+	TemplateBody string
+	Parameters   []*awsCF.Parameter
+}
```
Comment (xh3b4sd) [inline service/resource/cloudformation/resource.go]: FYI so far we put things like these into a `spec.go`.

Comment (fgimenez) [reply service/resource/cloudformation/resource.go]: Thx, moved there, I've also removed an unused field.
